src/Idris/IDEMode/CaseSplit.idr

```diff
@@ -65,69 +65,75 @@ isWhitespace : SourcePart -> Bool
 isWhitespace _              = False
 
+data Wrap = Brace | Auto | Paren | Bracket
+
 ||| Given a list of definitions, a list of mappings from `RawName` to `String`,
 ||| and a list of tokens to update, work out the updates to do, apply them, and
-||| return the result.
+||| return the result. Maintain a stack of brackets so we know if we need to insert
+||| a named application.
 doUpdates : {auto s : Ref Syn SyntaxInfo} ->
             {auto u : Ref UPD (List String)} ->
-            Defs -> Updates -> List SourcePart ->
+            Defs -> Updates -> List Wrap -> List SourcePart ->
             Core (List SourcePart)
-doUpdates defs ups [] = pure []   -- no more tokens to update, so we are done
+doUpdates defs ups stk [] = pure []   -- no more tokens to update, so we are done
 -- if we have a name that acts as an as-pattern then do not update it
-doUpdates defs ups (Name n :: AsPattern :: xs)
-    = pure $ Name n :: AsPattern :: !(doUpdates defs ups xs)
+doUpdates defs ups stk (Name n :: AsPattern :: xs)
+    = pure $ Name n :: AsPattern :: !(doUpdates defs ups stk xs)
 -- if we have an `@{` that was not handled as a named pattern, it should not
 -- result in named expansion (i.e. `@{n = ...}`) because that is not syntactically
--- valid. This clause allows us to treat `@{n}` just like `n` (no braces).
-doUpdates defs ups (AsPattern :: LBrace :: xs)
-    = pure $ AsPattern :: LBrace :: !(doUpdates defs ups xs)
--- if we have an LBrace (i.e. `{`), handle its contents
-doUpdates defs ups (LBrace :: xs)
-    -- the cases we care about are easy to detect w/o whitespace, so separate it
-    = let (ws, nws) = span isWhitespace xs in
+-- valid. Push `Auto` to mark this case.
+doUpdates defs ups stk (AsPattern :: LBrace :: xs)
+    = pure $ AsPattern :: LBrace :: !(doUpdates defs ups (Auto :: stk) xs)
+-- Handle tuples inside an implicit by pushing Paren
+doUpdates defs ups stk (Other "(" :: xs)  = pure (Other "(" :: !(doUpdates defs ups (Paren :: stk) xs))
+doUpdates defs ups stk (Other ")" :: xs)  = case stk of
+  (Paren :: stk) => pure (Other ")" :: !(doUpdates defs ups stk xs))
+  _ => pure (Other ")" :: !(doUpdates defs ups stk xs))
+doUpdates defs ups stk (Other "[" :: xs)  = pure (Other "[" :: !(doUpdates defs ups (Bracket :: stk) xs))
+doUpdates defs ups stk (Other "]" :: xs)  = case stk of
+  Bracket :: stk => pure (Other "]" :: !(doUpdates defs ups stk xs))
+  stk => pure (Other "]" :: !(doUpdates defs ups stk xs))
+doUpdates defs ups stk (Other "-" :: Other "-" :: xs) = pure (Other "--" :: xs)
+-- name after LBrace
+doUpdates defs ups stk (LBrace :: xs) =
+  let (ws, nws) = span isWhitespace xs in
+    case nws of
+      Name n :: rest => case lookup n ups of
+        Just up => pure (LBrace :: ws ++ Name n :: Whitespace " " :: Equal :: Whitespace " " :: Other up :: !(doUpdates defs ups stk rest))
+        Nothing => pure (LBrace :: !(doUpdates defs ups (Brace :: stk) xs))
+      _ => pure (LBrace :: !(doUpdates defs ups (Brace :: stk) xs))
```

Comment:
Can we not still jump past the whitespace in these cases?
Same remark applies later on too.